tonic-web: proxy any kind of Service

bouk opened the pull request:
This allows applying the GrpcWebLayer to any kind of Service, not just ones that tonic generates. This makes it possible to use tonic-web as a grpc-web proxy to a gRPC server implemented in another language for example.


## Motivation

See https://github.com/hyperium/tonic/issues/1361

## Solution

I made GrpcWebService generic over the request body it receives and the body that the upstream service responds. It's mostly a lot of search and replace with a couple of changes in how we convert the body to conform to the right interface.

I've also added a test that proxies to a gRPC service over a `hyper::Client` to show it's generic over any kind of Service.

Closes https://github.com/hyperium/tonic/issues/1361

LucioFranco commented on tonic-web/src/layer.rs:
```diff
@@ -4,32 +4,43 @@ use tower_layer::Layer;
+    }
+}
+
+impl<RespBody> GrpcWebLayer<RespBody> {
```
We tend to use `ResBody` instead of `RespBody`, so we should make that change everywhere here to follow hat we do in the rest of the project.

bouk commented on tonic-web/src/layer.rs:
```diff
@@ -4,32 +4,43 @@ use tower_layer::Layer;
 where
-    S: Service<http::Request<hyper::Body>, Response = http::Response<BoxBody>>,
-    S: Send + 'static,
+    S: Service<http::Request<BoxBody>, Response = http::Response<RespBody>> + Send + 'static,
```
This really broke my brain while working on it, but the BoxBody here is actually the Body that's forwarded from the GrpcWeb middleware to the upstream service. So it needs to be an actual type, because we don't just forward the body we get

bouk commented on tonic-web/src/service.rs:
```diff
@@ -121,7 +139,7 @@ where
                 ResponseFuture {
                     case: Case::Other {
-                        future: self.inner.call(req),
+                        future: self.inner.call(req.map(box_body)),
```
I think it's unavoidable because the middle are constructs its own bodies when it decodes the request

bouk commented:
@LucioFranco thanks for the quick review, I'll make the changes tomorrow

bouk commented:
@LucioFranco Please take another look

1. I've tried to keep it backwards-compatible by adding default type parameters
2. I un-exposed the axum IntoResponse trait by making the ResBody of the router a bit more constrained
3. I addressed some of your comments

bouk commented:
@LucioFranco could you take another look at this?

bouk commented:
Hi @LucioFranco I'm still interested in landing this, just rebased on master could you take another look?

bouk commented:
Since tonic has been upgraded to http v1 this will obviously need a rebase, which I'll do in a new PR :)
